3.2 proposed backports (June 2023 edition)

vadz opened the pull request.

TTJeGo committed:
Improve ScrollTo() when using wxDV_VARIABLE_LINE_HEIGHT

Calculate the last visible item based on item position instead of the
first visible row when using variable line heights.

Also extend the sample for testing EnsureVisible with item and column.

See #23102, #23128.

(cherry picked from commit 842ca1e8b8d754ed9782aa652d7cd00109848942)

vadz committed:
Fix menu width for owner-drawn items using bitmap in wxMSW

Do _not_ provide a valid HBITMAP for them, as this seems to break
measuring such items by not using our WM_MEASUREITEM handler at all.

Note that we could also use HBMMENU_CALLBACK for the bitmap, which also
fixes the problem, but it seems better to just not do anything at all.

See #23230.

Co-Authored-by: Jacob Gillespie <jacobgillespie@gmail.com>

(cherry picked from commit 1e98812a56a494dce60384ba01b7c0a77b4fd651)

vadz committed:
Draw borders for scrollable windows in wxGTK too

In many themes tree view window doesn't have a border, and so none of
our windows using wxHSCROLL and/or wxVSCROLL styles had it neither, even
if they didn't have wxNO_BORDER style.

Fix this by using the same border as is used for the entry widget (which
usually does have it) for all windows, not just those without scroll
styles.

In particular, this fixes the problem with the missing border for
wxListCtrl (which uses scrolling).

See #23236.

(cherry picked from commit 937b0688159c5a4a46e6e87fd70b87bd991d0a7c)

csomor committed:
Wrap wxStaticText contents under Mac too

Instead of just clipping the label, wrap it. This is more consistent
with wxMSW and wxGTK and also generally more useful.

See #23341, #23345.

(cherry picked from commit 9c8dfe5b2a8a6d80c25bee37e6a2d9d03656c3e7)

TcT2k committed:
Fix MSW lib names for ARM64 with CMake build

Any 64 platform had the added _x64 suffix. This will now be _arm64
when building for ARM64.

See #23347, #23355.

(cherry picked from commit 1edb38b54892273f2f932339d49b91f7b60fd3a4)

adrianlopezroche committed:
Fix wxPropertyGrid tab traversal under wxGTK

Tab traversal from a wxPropertyGrid didn't work across wxPanel
instances or when the property grid was followed by a static control
such as wxStaticText due to implementing it "manually".

Fix this by using Navigate() to implement this instead, and just keep a
SetFocus() hack which is still needed for wxGTK.

See #23354, #23358.

(cherry picked from commit 0331057abaff52a8850321b7890edca13ca93a22)

paulcor committed:
Fix GSource leak from idle processing

It is apparently not safe to call gtk_events_pending() from the idle
callback, as it can result in GLib failing to properly dereference some
GSource objects.

This leads to increasingly large amounts of CPU time being spent in GLib
processing internal lists of these GSource objects. So avoid calling
gtk_events_pending(), as it is not strictly necessary because our idle
source will remain installed if wxIdleEvent::RequestMore() is used, and
if no events have arrived the idle callback will be invoked again.

See #23364, #23368.

(cherry picked from commit 2d32b8cdd86b4d557340cdde49163226dc9d99c8)

vadz committed:
Fix using wxComboCtrl popup inside AUI floating panes

Handle correctly the strange case of the popup window being destroyed
by Windows itself, which happened when the AUI floating frame, which was
used as the parent of the popup when it was created for a wxComboCtrl
inside an AUI pane, was itself destroyed. This resulted in problems
because we kept using popup window even though its HWND wasn't valid any
longer.

Detect when this happens and destroy the popup window too, to force
recreating it when/if it's needed again in the future.

This also requires keeping wxOwnerDrawnComboBox::m_initChs even after
using them, as they will be needed again if we need to recreate the
popup.

See #23399.

(cherry picked from commit ac21107b9511fb8239f970d2e5e0a19301176742)

mcorino committed:
Fix wxRibbonToolBar::GetToolByPos() and DeleteToolByPos()

These functions didn't correctly account for the indices beyond the
first group.

Fix this and add tests for these functions to the sample.

See #23457, #23458.

(cherry picked from commit 38491111009c00714c21242c7d89616d4d77932d)

vadz committed:
Fix wxMSW wxWindow::GetTextExtent() with given font in high DPI

This function returned wrong results when using a non-default font in
non-standard DPI because it didn't adjust the font to use the window
DPI, as needs to be done manually in wxMSW before using it.

Fix this now and do the same thing here as we already did in
wxWindow::SetFont() (via WXAdjustFontToOwnPPI()) or in wxDC::SetFont().

See #23542.

(cherry picked from commit f3a4767455f62d1a614eac6d538599ac738d3918)

vadz committed:
Fix creating Chinese locales under macOS

Creating Chinese locales such as zh_TW didn't work because this string
didn't appear in [NSLocale availableLocaleIdentifiers] return value,
as only the variants explicitly specifying the script were included.

Fix this by ignoring the script part when checking if the locale is
available. This should work well for zh_TW or zh_SG, which exist only in
a single script (traditional and simplified, respectively), but it's not
really clear whether it works, or what should we do if it doesn't work,
for zh_CN or zh_HK both of which exist in both scripts.

See #23209.

(cherry picked from commit 6d849f8211977ab969f9fd4a503d6ad362f95e53)

utelle committed:
Fix parsing Windows sort order in wxLocaleIdent::FromTag()

Adjust the code to check for Windows sort order only after the last
underscore (instead of the first one).

This avoids misinterpreting locale tags like "zh_Hans_CN".

(cherry picked from commit 9ada6cd2db2191cdbe984efb86f0cba32d090861)

utelle committed:
Match the chosen script when creating wxUILocale under macOS

Take the script name into account if it's specified to avoid selecting a
locale with a different script.

(cherry picked from commit 9deb7ea74666a18d6ee0830996b039037f4fa2a4)

vadz committed:
Restore setting CRT locale in wxLocale under macOS

This was accidentally broken by the previous changes and the C-level
locale was not set at all any more, even when using wxLocale (and not
just wxUILocale, which is not supposed to set it) under macOS.

Fix this by explicitly calling setlocale() on this platform to restore
the previous, and documented, behaviour.

See #23557, #23603.

(cherry picked from commit 54a2eb16871e5159f2baa3228d0a6dd70b10633c)

BrianNixon committed:
Make wxSocket::Peek() work with UDP too

Fix wxSocketImpl::RecvDgram() when caller requests fewer bytes than
are present in datagram, as we must always read all the available bytes
when using UDP -- they are not buffered by the OS and so won't be
provided by the next Read() call.

See #23594, #23604.

(cherry picked from commit b9d0541f9ac4d4c356b6ab03772c1fe10a2fd1bc)

BrianNixon committed:
Clear grid attribute cache when loading new table

Clear the cached attributes when replacing the table they were used
with.

See #23607.

(cherry picked from commit 381c39a0518d5fc252eea05f69570e9ac767f536)

matyalatte committed:
Fix compile errors for wxNO_RTTI with some controls disabled

Fix compilation errors when wxNO_RTTI and wxUSE_UIACTIONSIMULATOR are
on, but one or more of wxUSE_COMBOBOX, wxUSE_CHOICE and wxUSE_LISTBOX
are off.

See #23611.

(cherry picked from commit 609b7e253551c17857c4303df1e9b0252cfeeccc)

PBfordev committed:
Copy all data files needed by XRC sample when building with CMake

XPM files demonstrating the use of wxBitmapBundle in the XRC files
were not copied when building the XRC sample with CMake.

This should have been part of eaa769a (Use wxBitmapBundle in
wxBitmapComboBox XRC handler, 2022-02-09).

See #23612.

(cherry picked from commit 241ed70b5bb643ba7c84e420d9cbba7618bed194)

MaartenBent committed:
CMake: Check and enable use of libxkbcommon

CMake's has built-in support for checking xkbcommon using FindX11 since
3.18.

But because older CMake versions are supported, add our own module that
checks if the header and library is available.

See #23410, #23613.

(cherry picked from commit a801057f1a7fba8e353be9481c1e621b8bf1dbe2)

TcT2k committed:
Include macOS 14 name in wxGetOsDescription()

Recognize macOS Sonoma too.

See #23615.

(cherry picked from commit db09adf6b5603cee96c404bf5bb95bc1e5901e75)

vadz committed:
Minor updates to wxUIActionSimulator documentation

Don't say that it's experimental in configure but do mention that it
doesn't work under Wayland.

Closes #23619.

(cherry picked from commit 58c49d06faf42e6968db6d1f0807e6a0dc5cd45c)

matyalatte committed:
Fix a linker error when wxUSE_PROTOCOL_FILE is off

Fix a linker error for wxProtocolUsewxFileProto when wxUSE_PROTOCOL_FILE
is off but wxUSE_PROTOCOL is on.

Closes #23629.

(cherry picked from commit c8fcb0be8ff74fef5e9f0efcbdd58414a79287cd)

dsa-t committed:
Fix wxTextCtrl size computation in wxGTK

Also improve wxTreeTextCtrl positioning in wxTreeCtrl.

See #23001, #23610, #23623.

(cherry picked from commit 27aa7e90a7882faeb9da42375e95ebba855ae1ca)

vadz committed:
Fix using generated wx-config in CMake builds

Since 1c64bd506c (Use grep instead of fgrep and egrep, 2023-05-10) EGREP
wasn't defined in CMake build any longer, so finding the correct config
script always failed.

Fix this by using "grep -E" directly instead of "$EGREP".

See #23537.

(cherry picked from commit 2b6be286ee10b963bd15b31313fe9bf89de5dbc2)

Kumazuma commented:
Can you add ff7476fcd0f3fccad19b849511b73cbfd9a317bd? It seem to be able backport.

MaartenBent committed:
CMake: Improve finding library name or path of imported targets

Checking LOCATION{_CONFIG} property on interfaces is only supported
since CMake 3.19. Also check the LOCATION and IMPORTED_LIBNAME
properties for possible library locations.

See #23632.

(cherry picked from commit 0f1f61a9c8c9fc8c4cb81a380df157e8171b0712)

Commit:
Avoid using invalid column index in wxOSX wxDataViewCtrl

We need to check for the column row to avoid generating an event with an
invalid index, which may result in a crash in the application code.

See #23636.

(cherry picked from commit 837382510d9e905783f3a2fc9acd84c8bddd0e1a)

Commit:
Avoid spurious clang "undeclared selector" warnings

We check that the selector is really supported before using it, so
suppress the unwanted warnings about doing this.

See #23639.

(cherry picked from commit f8af562b20e6d1d3ff70005177878bf62ecd59f2)

MaartenBent committed:
Only draw selection rectangle for selected items in wxVListBox

See #22562, #23641.

(cherry picked from commit 11ee5b4df574a4020f37f21fce43ae043245bb25)

PBfordev committed:
Improve documentation about parallel builds with MinGW makefile

In the note about building wxWidgets with MinGW makefile and using
parallel build, explain that "-jN" option cannot be used when building
"setup_h" target.

See #23642.

(cherry picked from commit 81ce7edb75ed799a2d4e91720501a089e3919fd4)

vadz committed:
Fix crash when using wxDataObjectComposite with bitmap in wxMSW

This reverts the changes of b1fad4da44 (Convert 0RGB wxBitmaps to RGB
when copying them to clipboard, 2017-04-30) that added a cast from
wxDataObject to wxBitmapDataObject for any object supporting bitmaps,
which resulted in a crash when this happened to be wxDataObjectComposite
and not wxBitmapDataObject, and does the same thing in
wxBitmapDataObject itself which now converts 0RGB bitmaps to RGB ones
when asked for the bitmap data.

The main change is that the code using wxDataObjectComposite doesn't
crash any longer, but another nice side effect is that this is possibly
more efficient as well because the conversion could be skipped entirely
if the bitmap is never pasted.

See #17640, #23564.

(cherry picked from commit 675728e1c085bd109abce67a99d0f8a1ab11bba0)

vadz committed:
Fix buffer overrun in wxHTMLDataObject under non-MSW platforms

Using strcpy() in GetDataHere() added an extra NUL at the end which
didn't fit into the buffer of the size returned by GetDataSize(). This
could have been also fixed by returning an extra byte from the latter
function, but as the string doesn't need to be NUL-terminated,
apparently, just use memcpy() with the correct number of bytes instead.

Also, because the string is not necessarily NUL-terminated, use the
provided length in wxHTMLDataObject::SetData() instead of relying on the
buffer being NUL-terminated and reading uninitialized memory beyond its
size.

See #17637, #23660.

Co-Authored-By: mcorino <martin@corino.nl>

vadz committed:
Fix memory leak of wxClipboard data on exit

When wxClipboard is destroyed as part of the program shutdown,
gdk_selection_owner_get() doesn't return our clipboard widget as owner
any more, so we don't reset the owner when Clear() is called and hence
never free the data.

Do it explicitly if we don't have clipboard ownership in Clear() any
longer to avoid memory leaks -- even though they are mostly harmless (as
they happen only once, on exit), they still show up in LSAN and similar
tools reports.

(cherry picked from commit b52728a62ae45a4f88126c5f4f6fafb04534daf5)

imciner2 committed:
Document already existing wxPrintPreview::GetZoom function

This base function was added over 16 years ago but never documented.

See #23667.

(cherry picked from commit 7669a7e1bdc79038c7a211158dab112da5b65d28)

konstantinbo committed:
Document wxDataViewCheckIconText

This will make wrapping it in wxPython easier.

See #23668, #23669.

(cherry picked from commit 691adc82418fb547b53c43c141ab1db7f2f25bde)

taler21 committed:
Fix bitmap returned from wxAuiToolBarItem::GetDisabledBitmap()

Return the disabled bitmap and not the normal one.

This fixes a regression introduced in 9a5123afed (Use wxBitmapBundle in
wxAuiToolBar and wxAuiPaneInfo too, 2021-11-14).

See #23666.

(cherry picked from commit eae24cc3b240a5728e94b805bb040b403fdfbe5c)

PBfordev committed:
Fix typos in changelog

Also add couple of missing words, while still keeping lines under 80
characters limit.

Closes #23675.

vadz committed:
Fix crash when reading truncated GIF image

Also add another check for EOF and possibly invalid "bits" value in
another place where we were not checking for it when reading from the
stream.

See #23409.

Co-authored-by: PB <PBfordev@gmail.com>

(cherry picked from commit c2e5749443bfee57ff1cd81af4dfbfb9e7c5480e)

vadz committed:
Relax checks on character entry in numeric validators

We can't perform the check for the invalid number in IsCharOk() because
it's perfectly normal for the number to be temporarily invalid while
it's being entered and, worse, sometimes this can't be avoided at all
and the existing behaviour prevented the user from entering _anything_
into a control limited to the values between 10 and 20.

Do ensure that the control still has a correct value after losing focus.

See #12968, #23561.

(combined cherry pick of c269932c4e8ed0c0e4a2554b928ae2b55f70ad8e,
dccddcdd59b9a11564b809c516b418356db21718 and
1c160557132f76423dbc0acb3a01b1f64f016dd1 from master)

vadz commented:
> Can you add [ff7476f](https://github.com/wxWidgets/wxWidgets/commit/ff7476fcd0f3fccad19b849511b73cbfd9a317bd)? It seem to be able backport.

Sorry, I'm not sure if it's really appropriate: this will change the default bitmap size from 16x15 to 16x16 which might be noticeable and it looks like a wrong thing to do in the stable branch. I could add `FromDIP()`, I guess, as this shouldn't break anything, would this be useful?

a-wi commented:
I think it would be good to backport 57a9247889ce7501321ef9b69171610582a482e3 - with `#define` (used in 3.2) instead of `constexpr`.

vadz commented:
@a-wi I don't think we can guarantee that the change of its numeric value doesn't break anything in the existing code, can we? E.g. just storing it (as a numeric value) in the config file and reading it back wouldn't work any more (yes, I agree that it's not the right thing to do, but people still do it and similar things all the time).

vadz committed:
Don't normalize the value of empty text control on focus loss

This is unexpected, as just TAB-bing in and out of the control changes
its value since the recent changes to wxNumValidator.

(cherry picked from commit 67f634dfc40fcf4b2a7c481c8c5b223c1494d1ba)

vadz committed:
Ignore keys with modifiers in wxNumValidatorBase

This allows to use accelerators, such as Alt-X, when the focus is on a
key control using numeric validator in wxGTK.

(cherry picked from commit 39045e1ce5a3f9c724fcdbb84d5377f5ba9c3a28)

vadz committed:
Move changelog entries accidentally added to 3.2.2 section

Somehow many entries got added to the wrong section and went under 3.2.2
instead of 3.2.2. Move them to the correct place now.

Also reintegrate the section for the changes in 3.2.2.1 which wasn't
present in this branch.

taler21 commented on docs/changes.txt:
```diff
@@ -271,6 +272,10 @@ wxGTK:
 - Fix loading WebKit2 extension when using wxWebView (Scott Talbert, #23497).
 - Fix editing cells without values in wxDataViewCtrl (#23523).
 - Fix uninitialized wxKeyEvent::m_isRepeat (jolz, #23593).
+- Also improve size computation of wxTextCtrl (Alex Shvartzkop, #23610).
```
"Also improve" should be replaced by just "Improve" because after moving the changelog entries to the correct place, there is no longer any context for "Also".

vadz replied:
Indeed, thanks for noticing this.

taler21 commented:
In the current changelog of the 3.2 branch, the entry for the hotfix version 3.2.2.1 is missing.
I think this should be added for completeness.

vadz commented on src/common/combocmn.cpp:
```diff
@@ -2367,6 +2407,14 @@ void wxComboCtrlBase::HidePopup(bool generateEvent)
 
+void wxComboCtrlBase::OnPopupDestroy()
+{
+    m_winPopup = nullptr;
+
+    delete m_popupInterface;
+    m_popupInterface = nullptr;
```
Yes, sorry, this slipped through. I really have to set up a local hook checking for this, to avoid running all the CI builds just to find this out...

vadz commented:
> In the current changelog of the 3.2 branch, the entry for the hotfix version 3.2.2.1 is missing. I think this should be added for completeness.

Indeed, thanks for noticing this (and for the review in general!), added now.

Kumazuma commented:
> I could add FromDIP(), I guess, as this shouldn't break anything, would this be useful?
Yes, it is.

vadz commented:
> > I could add FromDIP(), I guess, as this shouldn't break anything, would this be useful?
> 
> Yes, it is.

Done now in a2c6620907259ab006ada23d31446cf198e4c887.
